core/src/main/scala/org/apache/spark/io/CompressionCodec.scala

```diff
@@ -197,4 +201,8 @@ class ZStdCompressionCodec(conf: SparkConf) extends CompressionCodec {
     // avoid overhead excessive of JNI call while trying to uncompress small amount of data.
     new BufferedInputStream(new ZstdInputStream(s), bufferSize)
   }
+
+  override def zstdEventLogCompressedInputStream(s: InputStream): InputStream = {
+    new BufferedInputStream(new ZstdInputStream(s).setContinuous(true), bufferSize)
```

Comment:
What happens if you set continuous for everything? would it work in all cases? It kind of looks like zstd always uses this in the code path below anyway.

I think that if we introduce a new method we might try to make it a little more general, like: compressedInputStreamForPartialFile or something. It would be good to avoid the isInstanceOf below.

Reply:
Thanks @srowen for the review.
Since, the CompressionCodec class used by many classes, we need to see any use case for, whether to read open frame for zstd case. As far as the eventLoggingListener class is concerned, it needs the open frame data also. So, I tried to change as minimal as possible without impacting the other calls.


> I think that if we introduce a new method we might try to make it a little more general, like: compressedInputStreamForPartialFile or something. It would be good to avoid the isInstanceOf below.

Yeah. This is a cleaner solution. Thanks.

Reply:
Yeah, so this new "partial file" method can call the existing method by default, and do something different for zstd. Then this one call site can ask for the 'partial file' stream. Some comments about the difference here would be helpful.

Reply:
Thanks. I will update the PR.

Reply:
@srowen I updated the PR.

Reply:
BTW it seems like 'continuous' changes behavior very little: https://github.com/luben/zstd-jni/blob/master/src/main/java/com/github/luben/zstd/ZstdInputStream.java#L147 I agree with your concern to keep the change minimal. I'm trying to think if this would break anything if everything were read as 'continuous'. It wouldn't fail fast in some case?

Reply:
That's what I'm wondering about. Is it actually desirable to not fail on a partial frame? I'm not sure. We *shouldn't* encounter it elsewhere.

This changes a developer API, but may not even be a breaking change as there is a default implementation. We can take breaking changes in Spark 3 though.

I think I agree with your approach here in the end.

Reply:
Thanks @srowen . 

> Is it actually desirable to not fail on a partial frame? I'm not sure. We shouldn't encounter it elsewhere.
Yes. Ideally it shouldn't fail. Even for EventLoggingListener if the application is finished, the frame will close (That is why it is applicable for only running application). After analyzing again the zstd code, the impact seems lesser "Either throw exception or read the frame", and latter seems better.
I can update the code.

Reply:
I have updated the code.

Reply:
> Is it actually desirable to not fail on a partial frame?

If you're reading a shuffle file compressed with zstd, and the shuffle file is corrupted somehow, this change may be allowing Spark to read incomplete shuffle data...

Reply:
Thanks @vanzin for the review. 

It seems we need both the methods in the ZstdCompressionCodec class, like in the previous change.

Reply:
Updated the code